### DISCUSSION TITLE
Implementing persistence / checkpointers in streamlit UI

### DISCUSSION THREAD

**Discussion by sreyakumar** [JoshuaC215/agent-service-toolkit#112] [Q&A] [answered]
Hi,

I'm working with one langgraph model that I am able to successfully implement into streamlit. However, the model has no memory. The only memory related code I have is 
`memory = MemorySaver()
async_app = async_workflow.compile(checkpointer=memory)`

I was wondering if someone could explain the structure/logic behind how memory is implemented in your package? Like this portion -- 
`thread_id = get_script_run_ctx().session_id` .  

I've tried just using one arbitrary id number like 1 in my threads but that fails too. Thanks!

**Comment by JoshuaC215** [comment, marked as answer]
Hey, you probably need to store the MemorySaver() in st.session_state and re-use it across runs instead of declaring it as its own variable, since Streamlit will re-create it on each run with the code you shared.

Fair warning, you may run into some further issues. LangChain / LangGraph and Streamlit don't always play nice together when running in the same process. Hence why I split out the architecture in this repo.

I won't be able to provide more support on your use case but the Streamlit forums or LangGraph discussions might be a good place to ask further questions!
